src/combinatorics/adjacency_matrix.lean

```diff
@@ -0,0 +1,67 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author:  Aaron Anderson, Jalex Stark.
+-/
+import linear_algebra.matrix
+import data.rel
+import combinatorics.simple_graph
+
+open_locale classical big_operators
+open finset matrix simple_graph
+noncomputable theory
+
+universes u v
+variables {α : Type u} [fintype α] [decidable_eq α]
+variables (R : Type v) [semiring R]
+
+namespace simple_graph
+
+variables (G : simple_graph α) (R)
+
+/-- The matrix $A$ such that $A i j = 1$ if $i$ and $j$ are adjacent, and otherwise $A i j = 0$-/
+def adjacency_matrix : matrix α α R := λ i j, ite (G.adj i j) 1 0
+
+variable {R}
+
+theorem transpose_adjacency_matrix : (G.adjacency_matrix R).transpose = (G.adjacency_matrix R) :=
+by { ext, simp [adjacency_matrix, edge_symm], }
+
+@[simp]
+lemma adjacency_matrix_apply (v w : α) : G.adjacency_matrix R v w = ite (G.adj v w) 1 0 := rfl
+
+@[simp]
+lemma adjacency_matrix_dot_product (v : α) (vec : α → R) :
+  dot_product (G.adjacency_matrix R v) vec = ∑ u in G.neighbor_finset v, vec u :=
+by simp [neighbor_finset_eq_filter, dot_product, sum_filter]
+
+@[simp]
+lemma dot_product_adjacency_matrix (v : α) (vec : α → R) :
+  dot_product vec (G.adjacency_matrix R v) = ∑ u in G.neighbor_finset v, vec u :=
+by simp [neighbor_finset_eq_filter, dot_product, sum_filter, sum_apply]
+
+@[simp]
+lemma adjacency_matrix_mul_apply (M : matrix α α R) (v w : α) :
+  (G.adjacency_matrix R * M) v w = ∑ u in G.neighbor_finset v, M u w :=
+by { rw [mul_eq_mul, mul_val], simp [neighbor_finset_eq_filter, sum_filter], }
```

Comment:
Unless we are in a matrix ring, `matrix.mul` operator is preferred over `has_mul.mul`.
```suggestion
  (G.adjacency_matrix R ⬝ M) v w = ∑ u in G.neighbor_finset v, M u w :=
by simp [mul_val, neighbor_finset_eq_filter, sum_filter], }
```

(I thought the `simp` linter would notice this, but apparently not!)